typed-protocols/src/Network/TypedProtocol/Codec.hs

```diff
@@ -0,0 +1,148 @@
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE PolyKinds                  #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE NamedFieldPuns             #-}
+{-# LANGUAGE TypeInType                 #-}
+
+module Network.TypedProtocol.Codec (
+    Codec(..)
+  , PeerKind(..)
+  , PeerHasAgency(..)
+  , WeHaveAgency
+  , TheyHaveAgency
+  , SomeMessage(..)
+  , DecodeStep(..)
+  , transformCodec
+  , cborCodec
+  ) where
+
+import           Network.TypedProtocol.Core
+                   ( Protocol(..), PeerKind(..)
+                   , PeerHasAgency(..), WeHaveAgency, TheyHaveAgency )
+
+import           Control.Monad.ST (ST)
+import           Control.Monad.Class.MonadST
+
+import qualified Codec.CBOR.Encoding as CBOR (Encoding)
+import qualified Codec.CBOR.Read     as CBOR
+import qualified Codec.CBOR.Decoding as CBOR (Decoder)
+import qualified Codec.CBOR.Write    as CBOR
+
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Builder as BS
+import qualified Data.ByteString.Builder.Extra as BS
+import qualified Data.ByteString.Lazy.Internal as LBS (smallChunkSize)
+import qualified Data.ByteString.Lazy          as LBS
+
+data Codec ps (pk :: PeerKind) failure m bytes = Codec {
+       encode :: forall (st :: ps) (st' :: ps).
+                 WeHaveAgency pk st
+              -> Message ps st st'
+              -> bytes,
+
+       decode :: forall (st :: ps).
+                 TheyHaveAgency pk st
+              -> m (DecodeStep bytes failure m (SomeMessage st))
+     }
+
+transformCodec
+  :: Functor m
+  => (bytes  -> bytes')
+  -> (bytes' -> bytes)
+  -> Codec ps pk failure m bytes
+  -> Codec ps pk failure m bytes'
+transformCodec to from Codec {encode, decode} = Codec {
+    encode = \stok msg -> to (encode stok msg),
+    decode = fmap (transformDecodeStep to from) . decode
+  }
+
+-- The types here are pretty fancy. The decode is polymorphic in the protocol
+-- state, but only for kinds that are the same kind as the protocol state.
+-- The StateToken is a type family that resolves to a singleton, and the
+-- result uses existential types to hide the unknown type of the state we're
+-- transitioning to.
+--
+-- Both the Message and StateToken data families are indexed on the kind ps
+-- which is why it has to be a paramater here, otherwise these functions
+-- are unusable.
+
+
+data DecodeStep bytes failure m a =
```

Comment:
That's true, we decided the complexity / tightness trade-off was probably not worth it.

Reply:
I see no increase in complexity.

Reply:
Further, the tighter representation is still a monad.

Reply:
And I'd even say that the lack of tightness counts as complexity: everyone who deals with the loose decoder has to handle a case that doesn't make any sense.

Reply:
It's handled in one place.
```
runDecoder Channel{recv} = go
  where
    go _ (DecodeDone x trailing) = return (Right (x, trailing))
    go _ (DecodeFail failure)    = return (Left failure)
    go Nothing         (DecodePartial k) = recv >>= k        >>= go Nothing
    go (Just trailing) (DecodePartial k) = k (Just trailing) >>= go Nothing
```

Reply:
This is backwards. The channel is made to respond to the decoder, whereas it should be the decoder which responds to the channel: no matter what the channel gives, the decoder must decide `Either fail decoded`. Sure it's handled in one place here, but with this loose definition, an errant decoder can cause an infinite loop by always giving `DecodePartial`. Compare at something like this

```Haskell
newtype Scan i partial f total = Scan
  { runScan :: f (ScanChooses i partial f total)
  }

data ScanChooses i partial f total where
  ScanDone :: total -> ScanChooses i partial f total
  ScanMore :: ScanOffers i partial f total -> ScanChooses i partial f total

data ScanOffers i partial f total = ScanOffers
  { scanMore :: i -> Scan i partial f total
  , scanEnd  :: partial
  }

-- A Decoder which consumes no input sets the partial type to `forall x . x`.
-- A Decoder which never fails sets `fail ~ forall x . x`.
-- In practice, would probably use
--   `ExceptT fail`
-- so that users can write monadic decoders
-- (Scan token expect fail f is a monad when f is a monad)
-- This is only for decoders where the token type really is the token type, not
-- ByteString where Word8 is actually needed, and so it doesn't deal with
-- leftovers (there is no such thing), but that could be worked in I believe.
type Decoder token expect fail f t = Scan token expect f (Either fail t)

runDecoder (Channel{recv}) scan = runScan scan >>= go
  where
  go (ScanDone total) = pure (Right total)
  go (ScanMore offer) = recv >>= \mToken -> case mToken of
    Nothing -> pure (Left (scanEnd offer))
    Just token -> go (scanMore token)
```

It's just a rough sketch but it shows the benefit: whenever the channel gives `Nothing`, we always get a result. If `runDecoder` loops forever, it must be the channel which loops forever, which is what you would expect.

Reply:
And it should be possible to dual-ize this stuff to get an encoder. It even expresses send and receive channels.

```Haskell
newtype CoScan o partial f total = CoScan
  { runCoScan :: f (CoScanOffers o partial f total)
  }

data CoScanOffers o partial f total = CoScanOffers
  { coScanDone :: total
  , coScanMore :: CoScanChooses o partial f total
  }

data CoScanChooses o partial f total where
  CoScanDone :: partial -> CoScanChooses o partial f total
  CoScanMore :: o -> CoScan o partial f total -> CoScanChooses o partial f total

type Encoder token f = CoScan token () f ()

type Send chunk f = Scan chunk () f ()

type Recv chunk f = CoScan chunk () f ()

connect
  :: ( Monad f )
  => Scan chunk partialS f totalS
  -> CoScan chunk totalC f partialC
  -> f (Either (totalS, CoScanOffers chunk totalC f partialC)
               (totalC, ScanChooses chunk partialS f totalS)
       )
connect scan coscan = do
  scanChooses <- runScan scan
  coScanOffers <- runCoScan coscan
  case scanChooses of
    ScanDone totalS -> pure $ Left (totalS, coScanOffers)
    ScanMore scanOffers -> case coScanMore coScanOffers of
      CoScanDone totalC -> pure $ Right (totalC, scanChooses)
      CoScanMore chunk coScan' ->
        connect (scanMore scanOffers chunk) coScan'

connectSendRecv
  :: ( Monad f )
  => Send chunk f
  -> Recv chunk f
  -> f (Either ((), Recv chunk f) ((), Send chunk f))
connectSendRecv send recv = connect send recv >>= \outcome -> case outcome of
  Left  (u, recv') -> pure $ Left  (u, CoScan (pure recv'))
  Right (u, send') -> pure $ Right (u, Scan   (pure send'))

connectDecoderEncoder
  :: ( Monad f )
  => Decoder token expect fail f t
  -> Encoder token f
  -> f ( Either (Either fail t, Encoder token f)
                ((), Decoder token expect fail f t)
       )
connectDecoderEncoder decoder encoder = connect decoder encoder >>= \outcome -> case outcome of
  Left  (t, encoder') -> pure $ Left  (t, CoScan (pure encoder'))
  Right (t, decoder') -> pure $ Right (t, Scan   (pure decoder'))
```

Reply:
Just as you can connect a send channel to a receive channel, with the _very same mechanism_ you can connect an encoder to a send channel, or a decoder to a receive channel.

Reply:
OK one more snippet and then I'll stop. This is just too cool

```Haskell
import Control.Monad.Trans.Free
import Control.Comonad.Trans.Cofree

data ScanF i partial t where
  ScanF :: partial -> (i -> t) -> ScanF i partial t

instance Functor (ScanF i partial) where
  fmap f (ScanF p k) = ScanF p (f . k)

-- Whereas ScanF is a product, CoScanF is a sum.
-- Whereas the second componenet of ScanF is a function, the second summand of
-- CoScanF is a multiplicative conjuncation. These are dual in linear logic.
data CoScanF o total t where
  CoScanDoneF :: total -> CoScanF o total t
  CoScanMoreF :: o -> t -> CoScanF o total t

instance Functor (CoScanF o total) where
  fmap f term = case term of
    CoScanDoneF total -> CoScanDoneF total
    CoScanMoreF o next -> CoScanMoreF o (f next)

type Scan i partial f total = FreeT (ScanF i partial) f total

type CoScan o total f partial = CofreeT (CoScanF o total) f partial

-- The result of connecting complementary scan and coscan: either they both
-- finish simultaneously, or one finishes before the other.
data Connect t partialS partialC totalS totalC f where
  BothDone   :: totalS
             -> totalC
             -> Connect t partialS partialC totalS totalC f
  ScanDone   :: totalS
             -> partialC
             -> t
             -> CoScan t totalC f partialC
             -> Connect t partialS partialC totalS totalC f
  CoScanDone :: totalC
             -> partialS
             -> (t -> Scan t partialS f totalS)
             -> Connect t partialS partialC totalS totalC f

connect
  :: ( Monad f )
  => Scan   t partialS f totalS
  -> CoScan t totalC   f partialC
  -> f (Connect t partialS partialC totalS totalC f)
connect (FreeT fscan) (CofreeT fcoscan) = do
  freeF <- fscan
  partialC :< coScanF <- fcoscan
  case freeF of
    Pure totalS -> case coScanF of
      CoScanDoneF totalC -> pure $ BothDone totalS totalC
      CoScanMoreF o nextCoScan -> pure $ ScanDone totalS partialC o nextCoScan
    Free (ScanF partialS k) -> case coScanF of
      CoScanDoneF totalC -> pure $ CoScanDone totalC partialS k
      CoScanMoreF t nextCoScan -> connect (k t) nextCoScan
```

Taking the special case of an encoder as a `CoScan`, we can `extract :: (Encoder token s -> t) -> Encoder token s -> Encoder token t` which I can imagine would find some application in, say, tracking the number of bytes that the encoder wrote (this is the "partial" value of the encoder). The comonadic-ness allows for sequencing encoders such that the next encoder is aware of the final partial value of the previous encoder, allowing it to adjust its meaning depending on that: too many bytes? allocate another buffer.

Reply:
> It's handled in one place.

Until it's not. https://github.com/input-output-hk/ouroboros-network/pull/249/files#diff-83b90bbd642add21a4351e069869662cR141